kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/FilterInvoker.java

```diff
@@ -72,34 +74,40 @@ default boolean shouldHandleResponse(ApiKeys apiKey, short apiVersion) {
      * <p>Handle deserialized request data. It is implicit that the underlying filter
      * wants to handle this data because it indicated that with {@link #shouldHandleRequest(ApiKeys, short)}
      * </p><p>
-     * Most Filters will want to call a method on {@link KrpcFilterContext} like {@link KrpcFilterContext#forwardRequest(RequestHeaderData, ApiMessage)}
+     * Most Filters will want to call a method on {@link KrpcFilterContext} like {@link KrpcFilterContext#completedForwardRequest(RequestHeaderData, ApiMessage)}
      * so that the message continues to flow through the filter chain.
      * </p>
-     * @param apiKey the key of the message
-     * @param apiVersion the apiVersion of the message
-     * @param header the header of the message
-     * @param body the body of the message
+     *
+     * @param apiKey        the key of the message
+     * @param apiVersion    the apiVersion of the message
+     * @param header        the header of the message
+     * @param body          the body of the message
      * @param filterContext contains methods to continue the filter chain and other contextual data
+     * @return
      */
-    default void onRequest(ApiKeys apiKey, short apiVersion, RequestHeaderData header, ApiMessage body, KrpcFilterContext filterContext) {
-        filterContext.forwardRequest(header, body);
+    default CompletionStage<? extends FilterResult> onRequest(ApiKeys apiKey, short apiVersion, RequestHeaderData header, ApiMessage body,
```

Comment:
I'm wondering if we should have this return CompletionStage<RequestFilterContext> and enable the user to say: `context.requestFilterResultBuilder().shortCircuitResponse().withMessage(msg)`.

And on the response side it would return `CompletionStage<ResponseFilterContext>` and the user would have no way to create a `ResponseFilterContext` that forwards a request.

So we start distinguishing the request and response sides and make it clearer what actions can be taken.

Reply:
Oh yes, I like that idea much better.  

So I think the consequences are:

1. `onXxxxRequest` methods return `CompletionStage<RequestFilterResult>`
2. `RequestFilterResult` can optionally a `ResponseFilterResult` (which is the short circuit response).  The builder will need to take care of ensuring the `RequestFilterResult` has either a request or a short circuit response. 
3. `RequestFilterResult<REQ,RES>` and `ResponseFilterResult<RES>` now have generic parameters, meaning the API becomes type-safe.

Am I on the right track?

Reply:
I tried making the API typesafe but failed. Java doesn't permit covariance on generics, which is what we hit here:


```
# Filter API signatures look like:
CompletionStage<RequestFilterResult<ProduceRequestData>> io.kroxylicious.proxy.filter.ProduceRequestFilter#onProduceRequest```

```
# Filter Invoker looks like this:
CompletionStage<RequestFilterResult<ApiMessage>> io.kroxylicious.proxy.filter.FilterInvoker#onRequest
```
This leads to a compilation error in `ProduceRequestFilterInvoker` because of an attempt to return  CompletionStage<RequestFilterResult<ProduceRequestData>>  as CompletionStage<RequestFilterResult<ApiMessage>>.




io.kroxylicious.proxy.filter.FilterInvoker#onRequest
io.kroxylicious.proxy.filter.FilterInvoker#onRequest
caused a compilation error

Reply:
I had a go at adding in the type parameters https://github.com/k-wall/kroxylicious/pull/5

And yes it does mean you have to do something wonky to get between `X<ApiMessage>` and `X<SpecificType>`, either unchecked casts or creating new objects.